qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,864 @@
+
+            \{f(t+0.5) \in \mathbb{C} | t \in \mathbb{Z} \wedge  0<=t<\texttt{duration}\}
+
+        Returns:
+            A waveform representation of this pulse.
+
+        Raises:
+            PulseError: When parameters are not assigned.
+            PulseError: When expression for pulse envelope is not assigned.
+        """
+        if self.is_parameterized():
+            raise PulseError("Unassigned parameter exists. All parameters must be assigned.")
+
+        if self._envelope is None:
+            raise PulseError("Pulse envelope expression is not assigned.")
+
+        fargs = _get_expression_args(self._envelope, self.parameters)
+        return Waveform(samples=self._envelope_lam(*fargs), name=self.name)
+
+    def validate_parameters(self) -> None:
+        """Validate parameters.
+
+        Raises:
+            PulseError: If the parameters passed are not valid.
+        """
+        if self.is_parameterized():
+            return
+
+        if self._constraints is not None:
+            fargs = _get_expression_args(self._constraints, self.parameters)
+            if not bool(self._constraints_lam(*fargs)):
+                param_repr = ", ".join(f"{p}={v}" for p, v in self.parameters.items())
+                const_repr = str(self._constraints)
+                raise PulseError(
+                    f"Assigned parameters {param_repr} violate following constraint: {const_repr}."
+                )
+
+        if self.limit_amplitude:
+            if self._valid_amp_conditions is not None:
+                fargs = _get_expression_args(self._valid_amp_conditions, self.parameters)
+                check_full_waveform = not bool(self._valid_amp_conditions_lam(*fargs))
+            else:
+                check_full_waveform = True
+
+            if check_full_waveform:
+                # Check full waveform only when the condition is satisified or
+                # evaluation condition is not provided.
+                # This operation is slower due to overhead of 'get_waveform'.
+                if not _is_amplitude_valid(self):
+                    param_repr = ", ".join(f"{p}={v}" for p, v in self.parameters.items())
+                    raise PulseError(
+                        f"Maximum pulse amplitude norm exceeds 1.0 with parameters {param_repr}."
+                        "This can be overruled by setting Pulse.limit_amplitude."
+                    )
+
+    def is_parameterized(self) -> bool:
+        """Return True iff the instruction is parameterized."""
+        return any(isinstance(val, ParameterExpression) for val in self.parameters.values())
+
+    @property
+    def parameters(self) -> Dict[str, Any]:
+        params = {"duration": self.duration}
+        params.update(self._params)
+        return params
+
+    def __eq__(self, other: "SymbolicPulse") -> bool:
+
+        if not isinstance(other, SymbolicPulse):
+            return NotImplemented
+
+        if self._pulse_type != other._pulse_type:
```

Comment:
I think there are two ways `__eq__` could work. Either it could ensure that all data fields are equal or it could ensure that `get_waveform` would produce the same pulse for each instance. For the first intent, the method needs to also compare constraints and valid_amp_conditions (and name and limit_amplitude). For the second, I don't think `_pulse_type` should be checked.

Reply:
For now this check verifies the waveform equality and qobj equality, and later needs `pulse_type` check too. Perhaps this can be removed once the providers migrate into the QPY model.

Reply:
Okay, if you think equality and hashing are mainly needed for Qobj serialization and could be removed soon, perhaps the details are not that important. I don't think there is too much overhead to being more thorough with them. My main concern is that these are generic methods that users could try to use in other ways, so it would be better to make them the way one would expect not knowing about the implementation (so `__eq__` and `__hash__` would make use of all the data that make a pulse unique).

Reply:
I agree with the criteria. Since I added envelope to the hash, now these methods evaluate the same set of data. I still think the `pulse_type` is important because some backend (not limited to IBM) may want pulse-name based matching for preset pulses (e.g. we could dynamically generate common pulse in hardware in parametrized fashion). In this sense, the `pulse_type` is one of information to get identical pulse at actual hardware. Of course, we expect backend to use envelope function rather than type, so we must implicitly assume unique mapping of name to envelope function.